pkg/server/plugin/upstreamca/awssecret/awssecret.go

```diff
@@ -0,0 +1,201 @@
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/secretsmanager/secretsmanageriface"
+	"github.com/hashicorp/hcl"
+
+	"github.com/spiffe/spire/pkg/common/pemutil"
+	"github.com/spiffe/spire/pkg/common/x509svid"
+	"github.com/spiffe/spire/pkg/common/x509util"
+	spi "github.com/spiffe/spire/proto/common/plugin"
+	"github.com/spiffe/spire/proto/server/upstreamca"
+)
+
+const (
+	pluginName = "aws_secrets"
+)
+
+type AWSSecretConfiguration struct {
+	TTL             string `hcl:"ttl" json:"ttl"` // time to live for generated certs
+	Region          string `hcl:"region" json:"region"`
+	CertFileARN     string `hcl:"cert_file_arn" json:"cert_file_arn"`
+	KeyFileARN      string `hcl:"key_file_arn" json:"key_file_arn"`
+	AccessKeyID     string `hcl:"access_key_id" json:"access_key_id"`
+	SecretAccessKey string `hcl:"secret_access_key" json:"secret_access_key"`
+	SecurityToken   string `hcl:"secret_token" json:"secret_token"`
+}
+
+type awssecretPlugin struct {
+	serialNumber x509util.SerialNumber
+
+	mtx        sync.RWMutex
+	cert       *x509.Certificate
+	upstreamCA *x509svid.UpstreamCA
+
+	hooks struct {
+		getenv    func(string) string
+		newClient func(config *AWSSecretConfiguration, region string) (secretsmanageriface.SecretsManagerAPI, error)
+	}
+}
+
+func (m *awssecretPlugin) Configure(ctx context.Context, req *spi.ConfigureRequest) (*spi.ConfigureResponse, error) {
+	var (
+		sm  secretsmanageriface.SecretsManagerAPI
+		err error
+	)
+
+	config, ttl, err := m.validateConfig(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	// set the AWS configuration and reset clients +
+	// Set local vars from config struct
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	sm, err = m.hooks.newClient(config, config.Region)
+
+	if err != nil {
+		return nil, err
+	}
+
+	key, cert, err := fetchFromSecretsManager(ctx, config, sm)
+	if err != nil {
+		return nil, err
+	}
+
+	m.cert = cert
+	m.upstreamCA = x509svid.NewUpstreamCA(
+		x509util.NewMemoryKeypair(cert, key),
+		req.GlobalConfig.TrustDomain,
+		x509svid.UpstreamCAOptions{
+			SerialNumber: m.serialNumber,
+			TTL:          ttl,
+		})
+
+	return &spi.ConfigureResponse{}, nil
+}
+
+func (*awssecretPlugin) GetPluginInfo(context.Context, *spi.GetPluginInfoRequest) (*spi.GetPluginInfoResponse, error) {
+	return &spi.GetPluginInfoResponse{}, nil
+}
+
+func (m *awssecretPlugin) SubmitCSR(ctx context.Context, request *upstreamca.SubmitCSRRequest) (*upstreamca.SubmitCSRResponse, error) {
+	m.mtx.RLock()
+	defer m.mtx.RUnlock()
+
+	if m.upstreamCA == nil {
+		return nil, errors.New("invalid state: not configured")
+	}
+
+	cert, err := m.upstreamCA.SignCSR(ctx, request.Csr)
+	if err != nil {
+		return nil, err
+	}
+
+	return &upstreamca.SubmitCSRResponse{
+		SignedCertificate: &upstreamca.SignedCertificate{
+			CertChain: cert.Raw,
+			Bundle:    m.cert.Raw,
+		},
+	}, nil
+}
+
+func New(fn ...func(config *AWSSecretConfiguration, region string) (secretsmanageriface.SecretsManagerAPI, error)) upstreamca.Plugin {
```

Comment:
I'm so sorry for another round trip. I had a comment on this that I must have not posted.

I'd recommend not complicating the exported `New()` method to accomodate a test concern. I'd recommend either:
```
func New() upstreamca.Plugin {
    return newPlugin()
}

func newPlugin() *awssecretPlugin {
    p := &awssecretPlugin{
        serialNumber: x509util.NewSerialNumber(),
    }
    p.hooks.getenv = os.Getenv
    p.hooks.newClient = newSecretsManagerClient
    return p
```

The tests can then use `newPlugin()` instead of `New()` and overwrite the hooks on the returned value:
```
p := newPlugin()
p.hooks.newClient = newFakeSecretsManagerClient
```

Alternatively, `newPlugin()` takes the hook values as parameters:

```
p := newPlugin(newFakeSecretsManagerClient)
```